editor/editor/mode/visual.js

```diff
@@ -1,34 +1,105 @@
 /**
+ * External dependencies
+ */
+import classNames from 'classnames';
+
+/**
  * Internal dependencies
  */
 import InserterButton from '../../inserter/button';
 
-function Blocks( { blocks, onChange } ) {
-	const onChangeBlock = ( index ) => ( changes ) => {
-		const newBlock = {
-			...blocks[ index ],
-			changes
+class Blocks extends wp.element.Component {
+	constructor() {
+		super( ...arguments );
+		this.resetBlockState = this.resetBlockState.bind( this );
+		this.state = {
+			blocks: [],
+			selected: null,
+			hovered: null,
+		};
+	}
+
+	componentWillMount() {
+		this.setState( { blocks: this.props.blocks } );
+	}
+
+	componentWillReceiveProps( nextProps ) {
+		if ( nextProps.blocks !== this.state.blocks ) {
+			this.setState( { blocks: nextProps.blocks } );
+		}
+	}
+
+	onChangeBlock() {
+		return ( index ) => ( changes ) => {
+			const { blocks } = this.state;
+			const newBlock = {
+				...blocks[ index ],
+				attributes: {
+					...blocks[ index ].attributes,
+					...changes
+				}
+			};
+
+			const newBlocks = [
+				...blocks.slice( 0, index ),
+				newBlock,
+				...blocks.slice( index + 1 )
+			];
+
+			this.setState( { blocks: newBlocks } );
```

Comment:
It seems inevitable we'll need some concept of global state which can be accessed and updated from disparate components of the application. Redux is a good solution for this, but I find can be a pain point for new developer onboarding. If this were a path we'd go down, I'd want to make a conscious effort to keep it simple, ideally a single top-level reducer with a few obvious actions (maybe not even action creators or selectors).